core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -908,15 +921,37 @@ protected SearcherManager getSearcherManager() {
         return searcherManager;
     }
 
-    private Object dirtyLock(BytesRef uid) {
-        int hash = Murmur3HashFunction.hash(uid.bytes, uid.offset, uid.length);
-        return dirtyLocks[Math.floorMod(hash, dirtyLocks.length)];
+    private ReleasableLock dirtyLock(BytesRef uid) {
+        final Tuple<ReleasableLock, AbstractRefCounted> tuple = lockMap.compute(uid, (k, v) -> {
+            if (v == null || !v.v2().tryIncRef()) {
+                final ReleasableLock lock = locks.poll();
+                final AbstractRefCounted ref = new AbstractRefCounted(null) {
+                    @Override
+                    protected void closeInternal() {
+                        locks.offer(lock);
+                        lockMap.remove(k);
+                    }
+                };
+                return Tuple.tuple(lock, ref);
+            } else {
+                return v;
+            }
+        });
+        return tuple.v1();
     }
 
-    private Object dirtyLock(Term uid) {
+    private ReleasableLock dirtyLock(Term uid) {
         return dirtyLock(uid.bytes());
     }
 
+    private void remove(BytesRef uid) {
+        lockMap.get(uid).v2().decRef();
+    }
+
+    private void remove(Term uid) {
+        remove(uid.bytes());
+    }
+
```

Comment:
Cool, I wasn't aware of that implementation. I pushed e870b9de00dafacae5bdbd68777d56b991ac546b.